### DISCUSSION TITLE
Upload files as form-data over DDP

### DISCUSSION THREAD

**Discussion by diavrank** [meteor/meteor#11523] [Feature Requests]
I have read that exists a way to upload files using **meteor-files** package. However, many of us store files in the server or in a cloud storage but not in a database (where is the place that meteor-files package uses). Currently, if I want to upload files, for example an image, I have to **convert it to base 64 string** and after send this data over a meteor method. Another case is if I want to **upload multiple files** (which comes from an html form), I have to create a **REST endpoint** (I use `maka:rest` :) ) to accept files in **form-data**.

So, It would be great to have something like this:

```javascript
//Client
Meteor.callWithFormData('methodName',formData,callback);
```

I think that this feature would be revolutionary for **DDP protocol** to compete with **HTTP protocol**, and also **this would do development easier**. I am not saying that DDP could replace HTTP, but rather the idea would be to **use HTTP as a complement** for DDP (WebSockets) in real-time apps, since there are some parts where is mandatory to use HTTP, for example, when you have to implement **WebHooks** in other platforms to connect them with a meteor app, it is needed to publish REST endpoints since other platforms use HTTP clients. Another example are `<img src="url" />` tags where you put the URL of a resource, this need to be in HTTP.

AFAIK Meteor methods work in parallel for multiple clients (sessions). On the other hand, methods from a given client run one at a time, however, you can use **this.unblock()** to make them work in parallel. Based on this, if a big file is uploaded by a client, shouldn't affect to other clients.

**References:**

https://docs.meteor.com/api/methods.html#DDPCommon-MethodInvocation-unblock
https://forums.meteor.com/t/current-recommended-practice-for-this-unblock-in-subscriptions/35371/2

**Comment by JoCat** [comment]
Strongly agree, because of the need to use REST as a workaround, the meaning of having DDP is mostly lost.
